R/benchTradePerf.R

```diff
@@ -0,0 +1,214 @@
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell"}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
+#'      \item{\code{MktBench.*}: }{The benchmark and an arbitrary \code{type} provided as input (e.g. 'Open', 'Close')}
+#'      \item{\code{Performance}: }{The \emph{Benchmark PnL} performance, in bps}
+#' }
+#' 
+#' For \code{benchmark = 'VWAP'} it contains:
+#' \describe{
+#'      \item{\code{Symbol}: }{A string identifying the traded symbol to benchmark}
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell"}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
+#'      \item{\code{VWAP.*}: }{The benchmark and depending on \code{type} parameter either 'Txns' 'Mkt'}
+#'      \item{\code{Performance}: }{The \emph{VWAP PnL} metric, in bps}
+#' }
+#' 
+#' For \code{benchmark = 'PWP'} it contains:
+#' \describe{
+#'      \item{\code{Symbol}: }{A string identifying the traded symbol to benchmark}
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell"}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
+#'      \item{\code{POV}: }{The POV rate of the order}
+#'      \item{\code{PWP.Shares}: }{The ratio between the total unit traded and the POV rate}
+#'      \item{\code{PWP.Price}: }{Volume weighted price of the first \code{PWP.Shares} traded}
+#'      \item{\code{Performance}: }{The \emph{PWP PnL} metric, in bps}
+#' }
+#' 
+#' For \code{benchmark = 'RPM'} it contains:
+#' \describe{
+#'      \item{\code{Symbol}: }{A string identifying the traded symbol to benchmark}
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell".}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
+#'      \item{\code{t.Mkt.Volmn}: }{Total market volume over the order timespan}
+#'      \item{\code{t.Fav.Volmn}: }{Total market volume over the order timespan for which the average execution price of a 'Buy' ('Sell') order was lower (greater) than market prices}
+#'      \item{\code{t.Unfav.Volmn}: }{The opposite of \code{t.Fav.Volmn}}
+#'      \item{\code{RPM}: }{The \emph{relative performance measure}. Decimal in the 0 to 1 range.}
+#'      \item{\code{Quality}: }{A qualitiative RPM score over quintiles, bottom-up one of 'Poor', 'Fair', 'Average', 'Good', 'Excellent'. Present if \code{verbose = TRUE}}
+#' }
+#' 
+#' @seealso \code{\link{initPortf}}, \code{\link{addTxn}}
+#' 
+#' @examples 
+#' 
+#' @references Kissell, R. \emph{The Science of Algorithmic Trading and Portfolio Management} (ISBN 978-0-12-401689-7)
+#' 
+#' @author Vito Lestingi
+#' 
+#' @export
+#' 
+benchTradePerf <- function(Portfolio,
+                           Symbol,
+                           side = 1,
+                           benchmark = c("MktBench", "VWAP", "PWP", "RPM"),
+                           type = c("Txns", "Mkt"),
```

Comment:
I agree with you, there is little ambiguity around 'interval' and 'full' periods. Since this also points to the PWP arrival time, I suggest to gather this discussions. A couple of new params for dates may probably follow.

Reply:
Unfortunately i cannot do multi-line suggestions in reviewing a PR, so will post the full function code here. You can do a diff using BeyondCompare or some equivalent to see what is different, but i have simply added subsetting to interval.VWAP, besides replacing type=("Txns","Mkt") with type=("interval.VWAP","full.day.VWAP") in the function definition

PWP will require similar subsetting...but i will only be able to provide a suggestion this evening

`benchTradePerf <- function(Portfolio,
                           Symbol,
                           side = 1,
                           benchmark = c("MktBench", "VWAP", "PWP", "RPM"),
                           type = c("interval.VWAP", "full.day.VWAP"),
                           MktData,
                           POV = NULL,
                           verbose = FALSE
                           )
{ 
  pname <- Portfolio
  Portfolio <- .getPortfolio(pname)
  txns <- Portfolio[["symbols"]][[Symbol]][["txn"]]
  p_avg <- as.numeric(last(txns$Pos.Avg.Cost))
  tTxnQty <- sum(txns$Txn.Qty)
  
  # Benchmark metrics
  benchmark <- match.arg(benchmark, c("MktBench", "VWAP", "PWP", "RPM"), several.ok = TRUE)
  
  tradesPerf <- list()
  
  for (i in 1:length(benchmark)) {
    switch (benchmark[i],
            MktBench = {
              if (!("MktPrice" %in% colnames(MktData))) stop("No MktPrice column found, what did you call it?")
              if (missing(type)) type <- NULL
              
              benchPrice <- as.numeric(MktData[1, "MktPrice"])
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', paste(benchmark[i], type, sep = '.'))
            },
            VWAP = {# full VWAP
              if (missing(type)) type <- 'interval.VWAP'
              
              if (type == 'interval.VWAP') { # interval VWAP
                inds <- which(strftime(index(MktData), format = "%H:%M:%S", tz = "UTC") == strftime(index(txns), format = "%H:%M:%S", tz = "UTC"))
                inds <- sort(inds)
                start.idx <- first(inds)
                end.idx <- last(inds)
                benchPrice <- crossprod(MktData$MktPrice[start.idx:end.idx], MktData$MktVolmn[start.idx:end.idx])/sum(MktData$MktVolmn[start.idx:end.idx])
                } else if (type == 'full.day.VWAP') { # full.day.VWAP
                  if (!(("MktPrice" %in% colnames(MktData)) & ("MktVolmn" %in% colnames(MktData)))) stop("No MktPrice or MktVolmn column found, what did you call them?")
                  benchPrice <- crossprod(MktData[, "MktPrice"], MktData[, "MktVolmn"])/sum(MktData[, "MktVolmn"])
                }
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', paste(benchmark[i], type, sep = '.'))
            },
            PWP = {
              if (is.null(POV)) stop(paste("POV rate needed to compute", benchmark))
              # if (!("MktVolmn" %in% colnames(MktData))) stop("No MktVolmn column found, what did you call it?")
              # if (missing(arrTime)) arrTime <- first(index(txns))
              
              pwpShares <- rep(tTxnQty/POV, length(txns$Txn.Price))
              benchPrice  <- crossprod(txns$Txn.Price, pwpShares)/sum(pwpShares) # PWP price
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, POV, sum(pwpShares), benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 'POV', 'PWP.Shares', 'PWP.Price')
            },
            RPM = {
              if (!(("MktPrice" %in% colnames(MktData)) & ("MktVolmn" %in% colnames(MktData)))) stop("No MktPrice or MktVolmn column found, what did you call them?")
              
              tMktVolmn <- sum(MktData[, "MktVolmn"])
              
              if (side == 1) {
                tFavVolmn   <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] > p_avg])
                tUnfavVolmn <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] < p_avg])
              } else {
                tFavVolmn   <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] < p_avg])
                tUnfavVolmn <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] > p_avg])
              }
              
              rpm <- 0.5*(tMktVolmn + tFavVolmn - tUnfavVolmn)/tMktVolmn
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, tMktVolmn, tFavVolmn, tUnfavVolmn, rpm), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 't.Mkt.Volmn', 't.Fav.Volmn', 't.Unfav.Volmn', benchmark[i])
              
              # Append RPM qualitative score
              if (verbose) { 
                if (rpm >= 0.4 & rpm <= 0.6) {
                  quality <- "Average"
                } else if (rpm > 0.6) {
                  quality <- "Good"
                  if (rpm >= 0.8) {
                    quality <- "Excellent" 
                  }
                } else {
                  quality <- "Fair"
                  if (rpm < 0.2) {
                    quality <- "Poor"
                  }
                }
                out[, 'Quality'] <- quality 
              }
            }
    )
    
    # PnL performance for 'MktBench', 'VWAP' (all types) and 'PWP' 
    if (benchmark[i] != 'RPM') {
      benchPrice <- as.numeric(benchPrice)
      out[, 'Performance'] <- (-1) * side * (p_avg - benchPrice)/benchPrice * 10000
    }
    
    # Store and preserve data types
    tradesPerf[[i]] <- utils::type.convert(out, as.is = TRUE)
    names(tradesPerf)[i] <- paste('Trades', benchmark[i], 'Perf', sep = '.')
    row.names(tradesPerf[[i]]) <- NULL
  }
  return(tradesPerf)
}
`

Reply:
Thank you @jaymon0703. Started working on VWAP modifications.

* With your 'interval.VWAP' you introduced a different way to calculate the VWAP, which from the code is referred to market quantities as for your 'full.day.VWAP'. I found that VWAP is sometimes calculated with respect to transactional data as opposed to market one and `type` was used accordingly to account for this distinction and not for the timespan. If this is a correction, I would be thankful if you could please confirm and elaborate it a little more, as this is precisely the ambiguity I have been asking for clarification. 
* The coding style of variables you introduced is different from the camelCase requested. Assuming I should keep the latter.
* Why 'full.day'? There could be a very different timing, you did not change anything else in the `benchPrice`. May 'full' be more appropriate?

Waiting for PWP details, so I can commit these time-based modifications.

Reply:
- The volume weighted average price of all trades on the market *during* the time that your algo (or manual) order is trading is considered a fair price and is the general benchmark used when benchmarking execution performance with a VWAP benchmark. VWAP benchmarking means benchmarking your average execution price to interval VWAP (ie. the VWAP between the start and stop time of your order). If you compare your average execution to full day VWAP or VWAP to End of Day, you include an opportunity cost component of not trading.  
- intervalVWAP is fine but full.dayVWAP feels wrong. Google's styling code prefers camelCase with periods, but naming things is hard so whatever you feel comfortable with taking all this into account...as long as the function works i dont think it matters too much.
- the assumption is user has added full day's trades and Kissell refers to full day, which makes sense to compare the fair price for the entire day. If user does not have full day's trades, then full day VWAP will not be appropriate...and user will not use type="full day". This is why documentation will be impotant.

Reply:
* Time dimension acquired, the remaining question is: Is the VWAP always computed from market data, i.e. is it always the weighted average of market prices on market volume? 
* Pointed to `start.idx` and `end.idx`, but got your point... I will try to keep a consistent style if possible.
* Agree, but perhaps there is no need to include '.day'. If you input daily data you know that you are having a daily calc and the same goes for other timing. This usage of the function can be vignette examples?

Reply:
Regarding your question: "Time dimension acquired, the remaining question is: Is the VWAP always computed from market data, i.e. is it always the weighted average of market prices on market volume?"

The answer is yes.

I will resolve this conversation if you agree.

Reply:
Thank you @jaymon0703. Once the last PWP details on arrival time are discussed I can figure out if introducing a new time parameter of the function can suitably fit all of the above needs or if more of them are appropriate. Commenting about it in the review comment you opened.